packages/flutter_tools/lib/src/protocol_discovery.dart

```diff
@@ -49,50 +57,78 @@ class ProtocolDiscovery {
   final int devicePort;
   final bool ipv6;
 
-  final Completer<Uri> _completer = Completer<Uri>();
+  /// The time to wait before forwarding observatory URIs from [logReader].
+  final int throttleTimeInMilliseconds;
 
   StreamSubscription<String> _deviceLogSubscription;
 
+  StreamController<Uri> _uriStreamController;
+
   /// The discovered service URI.
+  /// Use [uris] instead.
+  // TODO(egarciad): replace `uri` for `uris`.
+  Future<Uri> get uri async {
+    try {
+      return await uris.first;
+    } on StateError {
+      // If the stream is closed and empty.
+      return null;
+    }
+  }
+
+  /// The discovered service URI stream.
   ///
   /// Port forwarding is only attempted when this is invoked, in case we never
   /// need to port forward.
-  Future<Uri> get uri async {
-    final Uri rawUri = await _completer.future;
-    return await _forwardPort(rawUri);
+  ///
+  /// Dependending on the lifespan of the app running the observatory,
+  /// a new observatory URI may be assigned to the app.
+  Stream<Uri> get uris {
+    return _uriStreamController.stream
+      .transform(_throttle<Uri>(
+        timeInMilliseconds: throttleTimeInMilliseconds,
+      ))
+      .asyncMap<Uri>((Uri observatoryUri) async {
```

Comment:
async/await is redundant:

```suggestion
      .asyncMap<Uri>((Uri observatoryUri) =>  _forwardPort(observatoryUri));
```